vm/src/compiler.rs

```diff
@@ -226,38 +213,40 @@ impl CompilerEnv for TypeInfos {
     }
 }
 
-pub struct Compiler<'a> {
-    globals: &'a (CompilerEnv + 'a),
+pub struct Compiler<'a, E> {
```

Comment:
Yeah, Rust does not downcast `&CompilerEnv` -> `&TypeEnv` automatically. An alternative solution may be to add an explict downcast method on `CompilerEnv`. 

```rust
// Could be a default method on CompilerEnv
fn as_type_env(&self) -> &TypeEnv { self }
```